core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -650,6 +653,28 @@ else if (redistributeWrites) {
                     partitioningScheme = Optional.of(new PartitioningScheme(Partitioning.create(FIXED_ARBITRARY_DISTRIBUTION, ImmutableList.of()), newSource.getNode().getOutputSymbols()));
                 }
             }
+            else if (scaleWriters
+                    && writerTarget.supportsReportingWrittenBytes(plannerContext.getMetadata(), session)
+                    && writerTarget.supportsMultipleWritersPerPartition(plannerContext.getMetadata(), session)
+                    // do not insert an exchange if partitioning is compatible
+                    && !newSource.getProperties().isCompatibleTablePartitioningWith(partitioningScheme.get().getPartitioning(), false, plannerContext.getMetadata(), session)) {
```

Comment:
It looks like `isCompatibleTablePartitioningWith` check does not change anything as you this check anyway in line 678. And only do anything if partitioning is not compatible.

Reply:
If not checking it here partitioning can get re-assigned and then the check at `678` won't pass